design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+
+#### Policy
+
+As with other authenticators, the SUT authenticator will be disabled by default.
+It will be enabled by creating an authenticator policy resource in Conjur and
+adding it to the list of enabled authenticators in the Conjur configuration, for
+example in the `CONJUR_AUTHENTICATORS` environment variable.
+
+The authenticator webservice must be declared in Conjur policy:
+
+```yml
+- !policy
+  id: conjur/authn-sut
+  body:
+  - !webservice
+
+  - !group 
+    id: authenticatable
+    annotations:
+      description: Group with permission to authenticate using this authenticator
+
+  - !permit
+    role: !group authenticatable
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+and the `authenticatable` group can be used to entitle a user(s) to use the
+authenticator:
+
+```yml
+- !grant
+  role: !group conjur/authn-sut/authenticatable
+  member: !user <username>
+```
+
+#### API
+
+We will need two new API endpoints for the SUT authenticator:
+
+##### `GET /authn-sut/<account>/login`
+
+This endpoint will be used to generate a SUT for a user. The user will provide
+their username and password (or API key) in the `Authentication` header using the
+standard format for HTTP Basic authentication.
+
+Additionally, the client will provide a Code Challenge and Code Challenge Algorithm
+in the headers. For example:
+
+```txt
+Code-Challenge: 3iod...
+Code-Challenge-Algorithm: sha256
+```
+
+For more details on the code challenge, see the [Code Verifier and Code Challenge](#code-verifier-and-code-challenge) section.
+
+The endpoint will return a SUT that can be used to authenticate the user to Conjur.
+
+The format of the URL matches the existing authenticator endpoint for
+authn-ldap.
+
+##### `POST /authn-sut/<account>/authenticate`
+
+This endpoint will be used to authenticate a user to Conjur using a SUT. The
+user will provide the SUT and code verifier in the request body.
+The endpoint will verify that the SUT is valid and has not expired, and that
+the code verifier matches the code challenge associated with the SUT. If this
+is the case, the endpoint will return a short-lived Conjur auth token.
+
+Example request body:
+
+```json
+{
+  "single_use_token": "lkduj...",
+  "code_verifier": "3iod..."
+}
+```
+
+The format of the URL matches the existing authenticator endpoint for
+authn-ldap.
+
+For more details on the code challenge, see the [Code Verifier and Code Challenge](#code-verifier-and-code-challenge) section.
+
+### Low Level Design
+
+#### Data Model
+
+The SUT will be a cryptographically secure, random string. It will be generated
+by the Conjur server and stored in the database. The SUT will be associated with
+the user and expiration time. The SUT will be deleted once it is utilized
+to authenticate the user to Conjur.
+
+When we store the SUT in the database, we will store a salted hash of the SUT instead
```

Comment:
Salting is a pattern to bring uniqueness to commonly used passwords. Essentially it removes the collision problem of hashed passwords. Given we'll be generating unique values to begin with, I don't believe the additional complexity of salting provides any additional security.

I'd recommend we take a similar approach for our single use tokens as we do when generating API keys:

```ruby
# Decomposed to expose the implementation details
require 'base32/crockford'

def random_token
  # note, this removes the down-casing to reduce collision potential
  OpenSSL::Random::random_bytes(32).unpack("N*").map{|i| Base32::Crockford::encode(i)}.join
end
```
If we store the above value hashed using SHA 256, there's essentially no way to get back to the original values even if the database was exposed.
I'd recommend we

Reply:
Good point. @andytinkham you originally [suggested](https://github.com/cyberark/conjur/pull/2971#discussion_r1349375507) salting. What do you say?

Reply:
As long as the token is of sufficient length, we should be OK not salting it.  Doing the same as the API keys should be OK.